Synchronized Pervasives implementation between stdlib and threads

mmottl opened the pull request:
While trying to add some other features, I noticed that the `Pervasives` implementations between the standard library and its threads implementation had diverged.  This patch merges the two implementations as much as possible and should make it easier to compare or diff them.

The only changes relevant to code execution are the following:

- `float_of_bits` in `threads` now uses the same external declaration as in `stdlib`.  There does not seem to be anything thread-specific there.
- `bytes_length` used the wrong primitive in `stdlib`.  It shouldn't really make a difference with the current implementation, but the patch makes it more future-proof.
- `flush_all` in the `stdlib` now only catches `Sys_error` exceptions.  I believe only assertion failures could be raised otherwise and might go unnoticed without the change.  Alternatively, we could also suppress all exceptions in `threads`, but maybe it's better to be more specific.
- `close_out` in `threads` suppressed exceptions escaping `flush`, which violates the manual and is inconsistent with `stdlib`.

Please let me know if you would prefer to merge the two implementations differently.

mmottl committed:
Synchronized Pervasives implementation between stdlib and threads

hcarty commented on stdlib/pervasives.ml:
```diff
@@ -369,7 +373,7 @@ external seek_out : out_channel -> int -> unit = "caml_ml_seek_out"
 external out_channel_length : out_channel -> int = "caml_ml_channel_size"
 external close_out_channel : out_channel -> unit = "caml_ml_close_channel"
-let close_out oc = flush oc; close_out_channel oc
+let close_out oc = (try flush oc with _ -> ()); close_out_channel oc
```
Buffered data could be silently lost if `flush` fails.  That makes this particular change seem dangerous.

mmottl replied:
@hcarty Note that this patch only makes `stdlib` and `threads` consistent.  In this case I chose the semantics as already implemented in `threads`, which I considered saner.  It will close the channel for sure, otherwise we would be leaking file descriptors.  If a user absolutely wants to make sure that their data has been flushed, they need to explicitly call `flush` themselves before calling `close_out`, thus potentially raising `Sys_error` exceptions.  `flush_all` also suppresses such exceptions, because the user could not possibly know anyway which channel was broken.

hcarty replied:
@mmottl Scope of a surrounding `try`/`with` would affect whether there is a leak or not here.

Regardless, this change should be reflected in the changelog and documentation to call out the requirement for an explicit call to `flush` to guarantee flushing of buffered output.

mmottl replied:
@hcarty Ok, I have looked at the documentation and it actually already specifies that `close_out` may raise even while flushing and before closing.  Users who don't like that behavior should use `close_out_noerr`, which completely ignores errors.  This means that the previous implementation in `threads` was not compliant with the manual while the `stdlib` was.

I have just submitted a patch to make the current `stdlib` behavior the default with both implementations.  They must definitely behave the same for that function.  It would be very confusing for users if that wasn't the case.  If some thread-related library needed the previous closing behavior, that code would have to be changed to use `close_out_noerr`.  I have looked, but haven't seen any such code in the distribution.  The semantics as specified in the manual should definitely prevail.

gasche commented:
I'm worried about changes that make `stdlib` follow `threads`'s behavior and not the other way around, as `threads`'s behavior is much more likely to be untested and not what people have been relying on. It may be the case that it is actually the better behavior (I suspect the `%bytes_length` vs. `%string_length` change is an aftermath of #596 and #772, but it looks like reverting to `string_length` here is an accidental change that was not required), but each would need really careful review -- maybe you should consider having separate PRs for these changes.

mmottl commented:
@gasche In this case I actually believe that it's the other way round: the `threads` behavior is much better tested.  I'm pretty sure that a majority of real-world applications, certainly I/O-heavy systems applications, links with the `threads` version.

Concerning `%bytes_length` vs `%string_length`, I'm sure that this is not in any way related to threads.  One of the two implementations must be the right one.  The type signature of the `external` declaration suggests that the `threads` version is the right one to use, otherwise I'd seriously question the naming convention used for the primitives.

mmottl commented:
I've just updated the initial comment describing the pull request to reflect changes wrt. `close_out`.

mshinwell commented:
I tend to agree that the threads library is probably in fact better tested.

gasche commented:
This may be true, but for all discrepancies pointed out above, `stdlib` is in fact the reference version. There is one case (`bytes_length`) where a stdlib change not reflected in threads was in fact a bad idea, and it's a good idea to revert it, but that does not change my comment that I was worried about changing "stdlib" to match "thread" without a very careful review.

ghost commented:
This patch touches otherlibs/threads, which is for vmthreads. I think none of the code in this directory is used when you are using system threads. I would think that no one uses vmthreads these days so the code in otherlibs/threads is likely to be the less tested one.

damiendoligez commented on stdlib/pervasives.ml:
```diff
@@ -521,7 +525,7 @@ external format_of_string :
  ('a, 'b, 'c, 'd, 'e, 'f) format6 = "%identity"
 
-let ( ^^ ) (Format (fmt1, str1)) (Format (fmt2, str2)) =
+let (^^) (Format (fmt1, str1)) (Format (fmt2, str2)) =
```
You should fix `threads/pervasives` to add the spaces rather than remove them here.

mmottl replied:
Sure, it is done.

mshinwell commented:
(I made a mistake in my previous comment above; I was thinking that the threads library being referred to was systhreads, but it is not.)

Is there a reason to keep the vmthreads library around?  Does anyone use it?

mmottl commented:
@mshinwell At the time the `vmthreads` library was implemented, POSIX threads were not yet that widely or well-supported by operating systems.  Xavier even wrote the first Linux implementation of POSIX threads (`LinuxThreads`), probably just so OCaml could have `systhreads` :smirk:.

I guess the question boils down to whether there are still any platforms left that don't have POSIX-threads and are worth supporting.  My guess is probably not.

mshinwell commented:
@xavierleroy Would you be happy removing vmthreads or not?

alainfrisch commented:
Does the implementation of vmthreads make sense in the context of compiling the ocaml bytecode interpreter + runtime system to WebAssembly, for instance?

xavierleroy commented:
You need to ask the greater OCaml user community to see if vmthreads are still in use.  

The only use I know of is in JoCaml, as a more lightweight alternative to system threads.  But JoCaml isn't particularly active these days...

Personally, it wouldn't break my heart to remove vmthreads altogether.  System threads are enough of a mess already :-)

ghost commented:
I can ask on the caml-list.

At worse I suppose that this could easily be split into a separate library. The only builtin support in the compiler seems to be adding `+vmthreads` to the search path when `-vmthreads` is passed. Moreover, this option is only useful for projects that don't use ocamlfind since ocamlfind already interprets -thread/-vmthreads and setups the search path accordingly.

@alainfrisch it seems that browsers are getting support for pthreads in javascript. I guess if this happens WebAssembly will get pthreads as well. In any case, lwt/async are much more popular and provides similar features, so it doesn't seem too bad to me to let anyone interested maintain it as a separate library.

BTW, we'll we need the consurtium approval to drop support for vmthreads?

alainfrisch reviewed (approved):
LGTM.

@mmottl Can you add a Changelog entry, marked as potentially breaking (mentioning the change of behavior for close_out in the thread version)?

mmottl commented:
@alainfrisch Thanks for the approval, Alain, and to all the reviewers.  I have added the Changelog entry and rebased + squashed this branch wrt. the current trunk.
